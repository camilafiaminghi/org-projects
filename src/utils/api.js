import axios from 'axios'

const RESTAPI_URL = 'https://api.github.com'
const RESTAPI_AUTH = 'token dd0f9fb05e1398f0784dd208d44d141b46134d5f'

// https://api.github.com/search/repositories?q=org:facebook+sort:stars&per_page=3
// https://api.github.com/repos/facebook/react/branches
// language:javascript
// axios.defaults.baseURL = process.env.REACT_APP_BASE_URL || "";
// axios.defaults.headers.common["Content-Type"]="application/json";
// axios.defaults.headers.common["Authorization"] = `Bearer${token}`;

export const getReposByOrg = ({org, sort, language, page}) => {
	/*
	 * sort by 'stars' || 'forks'
	 */
	const perPage = 4
	const withLanguage = (language) ? `+language:${language}`
	const api = `${RESTAPI_URL}/search/repositories?q=org:${org}+sort:${sort}${withLanguage}&per_page=${perPage}&page=${page}`
	return axios.get(api, {
		headers: {
			'Accept': 'application/json',
			'Authorization': RESTAPI_AUTH
		}
	})
	.then(response => response.data)
	.catch(error => { throw error })
}

export const getBranches = (action) => {
	const api = `${RESTAPI_URL}/repos/${action.org}/${action.repo}/branches`
	return axios.get(api, {
		headers: {
			'Accept': 'application/json',
			'Authorization': RESTAPI_AUTH
		}
	})
	.then(response => response.data)
	.catch(error => { throw error })
}
