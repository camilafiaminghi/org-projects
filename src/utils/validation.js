export function validationRules (name, value) {
	const rule = {
		search: !!value.trim() && value.length > 3
	}
	return rule[name]
}

export function isValid (form) {
	return Object.keys(form).every(function (key) {
		return form[key]
	});
}
