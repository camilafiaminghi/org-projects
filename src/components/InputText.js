import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { validationRules } from '../utils/validation'

class InputText extends Component {
	static propTypes = {
		name: PropTypes.string.isRequired,
		placeholder: PropTypes.string.isRequired,
		maxLength: PropTypes.number.isRequired,
		handleOnChange: PropTypes.func.isRequired,
		submitted: PropTypes.bool,
		charsLeft: PropTypes.bool,
		message: PropTypes.string,
		value: PropTypes.string
	}

	state = {
		text: '',
		changed: false,
		valid: false
	}

	handleValidation = (event) => {
		const value = event.target.value

		if ( value.length > 0 ) {
			const { name } = this.props
			const valid = validationRules(name, value)

			this.setState((prevState) => ({
				...prevState,
				text: value,
				changed: true,
				valid
			}))

			this.props.handleOnChange(name, value, valid)
		}
	}

	render() {
		const { changed, valid, text } = this.state
		const { name, placeholder, maxLength, charsLeft, message, submitted } = this.props
		const textLeft = maxLength - text.length

		return (
			<div>
				<input
					type="text"
					name={name}
					placeholder={placeholder}
					maxLength={maxLength}
					value={text}
					onChange={this.handleValidation}
					autoComplete="off" />

				{ (textLeft <= (maxLength - Math.round(maxLength/2)) && charsLeft && changed) && (<span>characteres left [{ textLeft }]</span>) }
				{ ((!valid && changed) && submitted) && (<span>{message}</span>) }
			</div>
		)
	}
}

export default InputText
