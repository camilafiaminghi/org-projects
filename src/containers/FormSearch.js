import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { getRepos } from './../actions/repos'
import InputText from './../components/InputText'
import { isValid } from '../utils/validation'

export class FormSearch extends Component {

	static propTypes = {
		handleRepos: PropTypes.func.isRequired,
		loading: PropTypes.bool
	}

	state = {
		form: {
			search: '',
		},
		validation: {
			search: false,
		},
		validated: false,
		submitted: false
	}

	handleOnChange = (name, value, valid) => {
		this.setState((prevState) => ({
			...prevState,
			form: {...prevState.form, [name]: value},
			validation: (prevState.validation.hasOwnProperty(name)) ? {...prevState.validation, [name]: valid} : {...prevState.validation}
		}))

		/* VALIDATE ENTIRE FORM */
		this.setState((prevState) => ({
			...prevState,
			validated: isValid(prevState.validation)
		}))
	}

	handleSubmit = (event) => {
		event.preventDefault()
		const { form } = this.state
		const { handleRepos } = this.props

		handleRepos(form.search)
	}

	render() {
		const { loading } = this.props
		const { validated, submitted, form } = this.state

		if ( loading ) {
			return (
				<div>Searching...</div>
			)
		}

		return (
			<div>
				<form onSubmit={this.handleSubmit}>
					<InputText
						name="search"
						placeholder="Organization name"
						maxLength={122}
						charsLeft={true}
						message="This field is required"
						handleOnChange={this.handleOnChange}
						submitted={submitted} />

					<button
						type="submit"
						disabled={!validated}
						className={validated ? 'btn' : 'btn disabled'}>
						Search
					</button>
				</form>
			</div>
		)
	}
}

export const mapStateToProps = ({ repos }) => {
	const { loading } = repos

	return {
		loading
	}
}

export const mapDispatchToProps = (dispatch) => {
  return {
    handleRepos: (org) => dispatch(getRepos(org))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(FormSearch);
