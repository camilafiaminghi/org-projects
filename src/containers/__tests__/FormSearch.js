import React from 'react';
import renderer from 'react-test-renderer'
import configureMockStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { FormSearch } from './../FormSearch'
import InputText from './../../components/InputText'
import { validation, isValid } from './../../utils/validation'

let provider
let wrapper

const mockStore = configureMockStore()
const store = mockStore({})
const props = {
	handleRepos: jest.fn()
}
const state = {
	form: {
		search: 'test',
	},
	validation: {
		search: true,
	},
	validated: true,
	submitted: false
}

describe('<FormSearch />', () => {

	beforeEach(() => {
		provider = shallow(<Provider store={store}><FormSearch {...props} /></Provider>)
		wrapper = provider.find(FormSearch).shallow()
	})

	it('matches snapshot', () => {
		const component = renderer.create(<FormSearch {...props} />)
		expect(component.toJSON()).toMatchSnapshot()
	})

	it('should render', () => {
		expect(wrapper.exists()).toBeTruthy()
	})

	it('should has InputText component', () => {
		expect(wrapper.find(InputText).exists()).toBeTruthy()
	})

	it('should has button type submit', () => {
		const button = wrapper.find('button')

		expect(button.exists()).toBeTruthy()
		expect(button.shallow().props().type).toEqual('submit')
		expect(button.shallow().props().disabled).toEqual(true)
	})

	it('should change state handleOnChange', () => {
		wrapper.instance().handleOnChange('search', 'test', true)
		expect(wrapper.state()).toEqual(state)
	})

	it('should enable button submit', () => {
		wrapper.setState(state)
		const button = wrapper.find('button')
		expect(wrapper.state()).toEqual(state)
		expect(button.shallow().props().disabled).toEqual(false)
	})

	it('should handleSubmit', () => {
		wrapper.setState(state)
		const form = wrapper.find('form')
		form.simulate('submit', {preventDefault() {}})
		expect(props.handleRepos).toHaveBeenCalledTimes(1)
	})
})
