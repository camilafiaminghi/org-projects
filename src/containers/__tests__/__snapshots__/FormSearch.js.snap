// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormSearch /> matches snapshot 1`] = `
<form
  className="row"
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    maxLength={122}
    name="search"
    onChange={[Function]}
    placeholder="Organization name"
    type="text"
    value=""
  />
  <button
    className="btn disabled"
    disabled={true}
    type="submit"
  >
    Search
  </button>
</form>
`;
